src/models/tree/build_tree.cu

```diff
@@ -45,31 +45,148 @@ struct NodeBatch {
 };
 
+class Histogram {
+ public:
+  using value_type = IntegerGPair;
+  // If we are using int64 as our type we need to do atomic adds as unsigned long long
+  using atomic_add_type = std::conditional<std::is_same_v<value_type::value_type, int64_t>,
+                                           unsigned long long,
+                                           value_type::value_type>::type;
+
+ private:
+  legate::Buffer<IntegerGPair, 3> buffer_;  // Nodes, outputs, bins
+  int node_begin_   = 0;
+  int node_end_     = 0;
+  std::size_t size_ = 0;
+
+ public:
+  Histogram(int node_begin, int node_end, int num_outputs, int num_bins, cudaStream_t stream)
+    : node_begin_(node_begin), node_end_(node_end)
+  {
+    static_assert(sizeof(IntegerGPair) == 2 * sizeof(atomic_add_type),
+                  "AtomicAdd type does not match size of type");
+    buffer_ =
+      legate::create_buffer<IntegerGPair, 3>({node_end - node_begin, num_outputs, num_bins});
+    size_ = (node_end - node_begin) * num_outputs * num_bins;
+    CHECK_CUDA(cudaMemsetAsync(
+      buffer_.ptr(legate::Point<3>::ZEROES()), 0, size_ * sizeof(IntegerGPair), stream));
+  }
+  Histogram() = default;
+
+  void Destroy()
+  {
+    if (size_ > 0) buffer_.destroy();
+    node_begin_ = 0;
+    node_end_   = 0;
+    size_       = 0;
+  }
+
+  bool ContainsBatch(int node_begin_idx, int node_end_idx)
+  {
+    return node_begin_idx >= node_begin_ && node_end_idx <= node_end_;
+  }
+
+  __device__ bool ContainsNode(int node_idx)
+  {
+    return node_idx >= node_begin_ && node_idx < node_end_;
+  }
+
+  IntegerGPair* Ptr(int node_idx) { return buffer_.ptr({node_idx - node_begin_, 0, 0}); }
+
+  std::size_t Size() { return size_; }
+
+  // Node, output, bin
+  __host__ __device__ IntegerGPair& operator[](legate::Point<3> p)
+  {
+    return buffer_[{p[0] - node_begin_, p[1], p[2]}];
+  }
+};
+
+class GradientQuantiser {
+  IntegerGPair scale;
+
+ public:
+  struct GetAbsGPair {
+    int num_outputs;
+    legate::AccessorRO<double, 3> g;
+    legate::AccessorRO<double, 3> h;
+    __device__ GPair operator()(int n) const
+    {
+      legate::Point<3> p = {n / num_outputs, 0, n % num_outputs};
+      return GPair{abs(g[p]), abs(h[p])};
+    }
+  };
+
+  // Calculate scale from upper bound on data
+  GradientQuantiser(legate::TaskContext context,
+                    legate::AccessorRO<double, 3> g,
+                    legate::AccessorRO<double, 3> h,
+                    legate::Rect<3> g_shape,
+                    cudaStream_t stream)
+  {
+    auto thrust_alloc = ThrustAllocator(legate::Memory::GPU_FB_MEM);
+    auto policy       = DEFAULT_POLICY(thrust_alloc).on(stream);
+    auto counting     = thrust::make_counting_iterator(0);
+    int num_outputs   = g_shape.hi[2] - g_shape.lo[2] + 1;
+    std::size_t n     = (g_shape.hi[0] - g_shape.lo[0] + 1) * num_outputs;
+    auto zip_gpair    = thrust::make_transform_iterator(counting, GetAbsGPair{num_outputs, g, h});
```

Comment:
We could and it would save a few bits of precision maybe. I decided this was easier for now but maybe if we want to try using 32 bit ints it could be helpful.